### DISCUSSION TITLE
Can I Push messages from IOT devices using TCP?

### DISCUSSION THREAD

**Discussion by AamerHejazi** [emqx/emqx#12188] [Q&A] [answered]
Dears,
I hope you are doing well,

Can I push data directly from any IOT device without mentioning the TOPIC name using TCP protocol (without the Flespi tool)?

Please advise me and suggest or give me ideas if that is possible.

Best Regards,
Amer Hijazi,

**Comment by thalesmg** [comment]
Hi.

I'm not entirely sure what you mean by "using TCP protocol" and "Flespi tool", but the MQTT protocol requires the topic to be specified when sending a `PUBLISH` packet.  So: no, it's not possible to publish without specifying a topic.

**Reply by AamerHejazi** [reply]
Hello @thalesmg 

Thank you for your response.
I mean using WebSocket over TCP for IOT tracker devices and this is the flespi website (https://flespi.com/).

Best Regards,
Amer Hijazi,

**Reply by thalesmg** [reply]
Hi.
The [websocket listeners](https://docs.emqx.com/en/enterprise/v5.3/configuration/listener.html#configure-websocket-listener) also follow the MQTT protocol, so one needs to specify the topic when using them as well.

**Reply by AamerHejazi** [reply]
I opened the discussion here by mistake

https://github.com/emqx/MQTTX/discussions/1520#discussioncomment-7888087

you will get a big pic about what I mean

**Reply by id** [reply, marked as answer]
@AamerHejazi EMQX is shipped with built-in Multi-Protocol Gateway which supports several non-MQTT protocols common in IoT and industrial world - take a look, may be some of them are supported by your devices. EMQX also has [ExProto](https://docs.emqx.com/en/enterprise/latest/gateway/exproto.html) gateway which can be used to support any other protocol assuming you could implement a service to parse events and call gRPC endpoint on EMQX side.

**Reply by thalesmg** [reply]
If you want to interface with an MQTT system, you'll need to specify a topic for each message according to your business rules, I'm afraid.

Some ideas to look into:
- Write your own server that is capable of assigning a topic to arriving messages and publishing them using MQTT to EMQX.
- Have a look at our [existing gateways](https://docs.emqx.com/en/enterprise/v5.3/gateway/gateway.html) and see if any of those are relevant.
- Have a look at [Neuron](https://github.com/emqx/neuron) and see if it's relevant to your use case (I'm not very familiar with it myself, but it might be close to what you're trying to achieve).

**Reply by AamerHejazi** [reply]
Thank you @id and @thalesmg for your kind answers and support
